airflow/utils/types.py

```diff
@@ -16,8 +16,31 @@
 # under the License.
 import enum
 
+from sqlalchemy.types import TypeDecorator, String
 
-class DagRunType(str, enum.Enum):
+
+class EnumString(TypeDecorator):
+    """
+    Declare db column with this type to make the column compatible with string
+    and string based enum values when building the sqlalchemy ORM query. It can
+    be used just like sqlalchemy.types.String, for example:
+
+    ```
+    class Table(Base):
+        __tablename__ = "t"
+        run_type = Column(EnumString(50), nullable=False)
+    ```
+    """
+    impl = String
+
+    def process_bind_param(self, value, dialect):
+        if isinstance(value, enum.Enum):
+            return value.value
+        else:
+            return value
+
+
+class DagRunType(enum.Enum):
```

Comment:
@ashb i removed the str here to avoid this type being misused in the rest of the code base. For example, if `DagRunType.SCHEDULED` is used for column types that are not declared as `EnumString`, developer will get the incorrect behavior regardless of which db driver is being used, so we can surface the issue sooner. That way, it's less likely to get half working PR merged into master.

Reply:
Not a fan of this - these enums are used elsewhere than just db.

Can you explain when this is a problem? Cos it worked for me in mysql

Reply:
How did you perform the mysql test? You should be able to reproduce the problem with my newly added test case in breeze. 

I am open to better way to handle this as well, but I think whatever solution we end with, we should aim for consistent behavior between db drivers.

Reply:
Yeah, I never tested upgrading an install using mysql, nor looked in the (mysql) db so missed it.

With your custom EnumString type can't we leave the enum subclass inheriting from `str` too?

Reply:
The downside of subclassing `str` is it will lead to inconsistent behavior between db drivers for columns that are not declared as `EnumString` type. My main concern is developer adding new db columns that are supposed to interact with python enum types, but declared it as `String` instead of `EnumString`. In this case, postgres users won't experience any problem, but mysql users will end up with crashes or corrupted data in database.

What's your main concern around lack of str subclassing? Are you worried that developer will get confused between use of `DagRunType.SCHEDULED.value` and `DagRunType.SCHEDULED`?

Reply:
Another way to address this problem is to fix it at db driver layer. i.e. changing mysqlclient to support enum types. If we can get this change merged upstream, then we can at least safely use str subclassed enum type for both postgres and mysql everywhere, but since the conversion is not done within sqlalchemy anymore, we need to double check other db drivers.

However, adding enum support to mysqlclient will take a bit more time. If we don't need to push this fix out sooner, then we can probably wait for that.

Or we can go with custom `EnumString` as a short term fix, then work on mysqlclient path as a follow up.

Reply:
Is there any reason why we do not want to use SqlAlchemy Enum Type? https://docs.sqlalchemy.org/en/13/core/type_basics.html#sqlalchemy.types.Enum 

I believe it will map everything correctly in all dialects

Reply:
i had the same question initially, i am curious about the reason for picking string for column type here. as far as i know, the biggest pain with using enum in postgres was schema evolution, i.e. adding new value to enum. I have heard postgres 9.x has made some improvements around that area.

Reply:
I pushed a new commit to switch to sqlalchemy enum type with native enum and check constraint disabled, which should address all our concerns without requiring a schema migration. I think this strikes a good balance.

@ashb @potiuk @kaxil thoughts? Or do you all prefer to go all in and migrate to native db enum type instead?